File: src/main/java/org/o3project/odenos/core/component/SystemManagerInterface.java
/*
 * Copyright 2015 NEC Corporation.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.o3project.odenos.core.component;

import org.o3project.odenos.core.manager.system.ComponentConnection;
import org.o3project.odenos.remoteobject.ObjectProperty;
import org.o3project.odenos.remoteobject.ObjectSettings;
import org.o3project.odenos.remoteobject.manager.ComponentTypesHash;
import org.o3project.odenos.remoteobject.manager.ObjectPropertiesHash;
import org.o3project.odenos.remoteobject.manager.ObjectPropertyList;
import org.o3project.odenos.remoteobject.message.Request;
import org.o3project.odenos.remoteobject.message.Response;
import org.o3project.odenos.remoteobject.messagingclient.MessageDispatcher;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.Map;

/**
 * SystemManagerInterface class.
 *
 */
public class SystemManagerInterface {
  private static final Logger log = LoggerFactory.getLogger(SystemManagerInterface.class);

  public static final String PROPETY_PATH = "property";
  public static final String SETTINGS_PATH = "settings";
  public static final String COMP_MNGS_PATH = "component_managers";
  public static final String COMP_MNG_PATH = "component_managers/%s";
  public static final String EVENT_MNGS_PATH = "event_managers";
  public static final String EVENT_MNG_PATH = "event_managers/%s";
  public static final String COMP_TYPES_PATH = "component_types";
  public static final String COMP_TYPE_PATH = "component_types/%s";
  public static final String COMPS_PATH = "components";
  public static final String COMP_PATH = "components/%s";
  public static final String CONNECTIONS_PATH = "connections";
  public static final String CONNECTION_PATH = "connections/%s";
  public static final String OBJECT_PATH = "objects/%s";

  private MessageDispatcher dispatcher;
  private String sourceObjectId = null;

  /**
   * Constructor.
   * @param dispatcher Message Dispatcher object.
   */
  @Deprecated
  public SystemManagerInterface(
      final MessageDispatcher dispatcher) {
    this.dispatcher = dispatcher;
    log.debug("Create SystemManagerInterface : Id = '"
        + this.getSystemManagerId() + "'.");
  }

  /**
   * Constructor.
   * @param dispatcher Message Dispatcher object.
   */
  public SystemManagerInterface(
      final MessageDispatcher dispatcher,
      final String sourceObjectId) {
    this.dispatcher = dispatcher;
    this.sourceObjectId = sourceObjectId;
    log.debug("Create SystemManagerInterface : Id = '"
        + this.getSystemManagerId() + "'.");
  }
  
  /**
   * Returns a system manager ID.
   * @return value of the system manager ID.
   */
  public String getSystemManagerId() {
    if (this.dispatcher == null) {
      return null;
    }
    return this.dispatcher.getSystemManagerId();
  }

  /**
   * Returns a Message Dispatcher.
   * @return Message Dispatcher.
   */
  public MessageDispatcher getDispatcher() {
    return dispatcher;
  }

  /**
   * Sets a Message Dispatcher.
   * @param dispatcher Message Dispatcher.
   */
  public void setDispatcher(MessageDispatcher dispatcher) {
    this.dispatcher = dispatcher;
  }

  // //////////////////////////////////////
  //
  // Basic Request
  //
  // //////////////////////////////////////

  /**
   * Requests a "GET Property".
   * @return value of the property.
   */
  public final ObjectProperty getProperty() {
    String path = PROPETY_PATH;
    log.debug("");
    Response resp = getObjectToSystemMng(path);
    if (resp == null) {
      return null;
    }
    try {
      return resp.getBody(ObjectProperty.class);
    } catch (Exception e) {
      log.error("Recieved Message Exception.", e);
      return null;
    }
  }

  /**
   * Requests a "GET Settings".
   * @return value of the settings.
   */
  public final ObjectSettings getSettings() {
    String path = SETTINGS_PATH;
    log.debug("");
    Response resp = getObjectToSystemMng(path);
    if (resp == null) {
      return null;
    }
    try {
      return resp.getBody(ObjectSettings.class);
    } catch (Exception e) {
      log.error("Recieved Message Exception.", e);
      return null;
    }
  }

  /**
   * Requests a "GET Component Managers".
   * @return value of the properties.
   */
  public final ObjectPropertyList getComponentMngs() {
    String path = COMP_MNGS_PATH;
    log.debug("");
    Response resp = getObjectToSystemMng(path);
    if (resp == null) {
      return null;
    }
    try {
      return resp.getBody(ObjectPropertyList.class);
    } catch (Exception e) {
      log.error("Recieved Message Exception.", e);
      return null;
    }
  }

  /**
   * Requests a "GET Event Managers".
   * @return value of properties.
   */
  public final ObjectProperty getEventMngs() {
    String path = EVENT_MNGS_PATH;
    log.debug("");
    Response resp = getObjectToSystemMng(path);
    if (resp == null) {
      return null;
    }
    try {
      return resp.getBody(ObjectProperty.class);
    } catch (Exception e) {
      log.error("Recieved Message Exception.", e);
      return null;
    }
  }

  /**
   * Requests a "GET ComponentTypes".
   * @return value of types.
   */
  public final ComponentTypesHash getComponentTypes() {
    String path = COMP_TYPES_PATH;
    log.debug("");
    Response resp = getObjectToSystemMng(path);
    if (resp == null) {
      return null;
    }
    try {
      return resp.getBody(ComponentTypesHash.class);
    } catch (Exception e) {
      log.error("Recieved Message Exception.", e);
      return null;
    }
  }

  /**
   * Requests a "GET Components".
   * @return value of properties.
   */
  public final ObjectPropertiesHash getComponents() {
    String path = COMPS_PATH;
    log.debug("");
    Response resp = getObjectToSystemMng(path);
    if (resp == null) {
      return null;
    }
    try {
      return resp.getBody(ObjectPropertiesHash.class);
    } catch (Exception e) {
      log.error("Recieved Message Exception.", e);
      return null;
    }
  }

  /**
   * Requests a "GET Connections".
   * @return map of connections.
   */
  public final Map<String, ComponentConnection> getConnections() {
    String path = CONNECTIONS_PATH;
    log.debug("");
    Response resp = getObjectToSystemMng(path);
    if (resp == null) {
      return null;
    }
    try {
      return resp.getBodyAsMap(ComponentConnection.class);
    } catch (Exception e) {
      log.error("Recieved Message Exception.", e);
      return null;
    }
  }

  /**
   * Requests a "GET Component Manager".
   * @param compMngId component manager ID
   * @return value of properties.
   */
  public final ObjectProperty getComponentManager(
      final String compMngId) {
    String path = String.format(COMP_MNG_PATH, compMngId);
    log.debug("");
    Response resp = getObjectToSystemMng(path);
    if (resp == null) {
      return null;
    }
    try {
      return resp.getBody(ObjectProperty.class);
    } catch (Exception e) {
      log.error("Recieved Message Exception.", e);
      return null;
    }
  }

  /**
   * Requests a "GET ComponentType".
   * @param compType component type.
   * @return value of properties.
   */
  public final ObjectPropertyList getComponentType(
      final String compType) {
    String path = String.format(COMP_TYPE_PATH, compType);
    log.debug("");
    Response resp = getObjectToSystemMng(path);
    if (resp == null) {
      return null;
    }
    try {
      return resp.getBody(ObjectPropertyList.class);
    } catch (Exception e) {
      log.error("Recieved Message Exception.", e);
      return null;
    }
  }

  /**
   * Requests a "GET Component".
   * @param compId component ID.
   * @return value of properties.
   */
  public final ObjectProperty getComponent(
      final String compId) {
    String path = String.format(COMP_PATH, compId);
    log.debug("");
    Response resp = getObjectToSystemMng(path);
    if (resp == null) {
      return null;
    }
    try {
      return resp.getBody(ObjectProperty.class);
    } catch (Exception e) {
      log.error("Recieved Message Exception.", e);
      return null;
    }
  }

  /**
   * Requests a "GET Connection".
   * @param connId connection ID.
   * @return value of connection.
   */
  public final ComponentConnection getConnection(
      final String connId) {
    String path = String.format(CONNECTION_PATH, connId);
    log.debug("");
    Response resp = getObjectToSystemMng(path);
    if (resp == null) {
      return null;
    }
    try {
      return resp.getBody(ComponentConnection.class);
    } catch (Exception e) {
      log.error("Recieved Message Exception.", e);
      return null;
    }
  }

  /**
   * Requests a "GET Object".
   * @param objId Object ID.
   * @return value of properties.
   */
  public final ObjectProperty getObject(
      final String objId) {
    String path = String.format(OBJECT_PATH, objId);
    log.debug("");
    Response resp = getObjectToSystemMng(path);
    if (resp == null) {
      return null;
    }
    try {
      return resp.getBody(ObjectProperty.class);
    } catch (Exception e) {
      log.error("Recieved Message Exception.", e);
      return null;
    }
  }

  /**
   * Requests a "PUT Property".
   * @param body a property object.
   * @return response object.
   */
  public final Response putProperty(
      final ObjectProperty body) {
    String path = PROPETY_PATH;
    log.debug("");
    return putObjectToSystemMng(path, body);
  }

  /**
   * Requests a "PUT Settings".
   * @param body a setting object.
   * @return response object.
   */
  public final Response putSettings(
      final ObjectSettings body) {
    String path = SETTINGS_PATH;
    log.debug("");
    return putObjectToSystemMng(path, body);
  }

  /**
   * Requests a "PUT Connection".
   * @param body a connection.
   * @return response object.
   */
  public final Response putConnection(
      final ComponentConnection body) {
    if (body == null) {
      return new Response(Response.BAD_REQUEST, null);
    }
    String path = String.format(CONNECTION_PATH, body.getObjectId());
    log.debug("");
    return putObjectToSystemMng(path, body);
  }

  /**
   * Requests a "PUT Component Manager".
   * @param body a property object.
   * @return response object.
   */
  public final Response putComponentMng(
      final ObjectProperty body) {
    String path = String.format(COMP_MNG_PATH, body.getObjectId());
    log.debug("");
    return putObjectToSystemMng(path, body);
  }

  /**
   * Requests a "PUT Component".
   * @param body a property object.
   * @return response object.
   */
  public final Response putComponent(
      final ObjectProperty body) {
    String path = String.format(COMP_PATH, body.getObjectId());
    log.debug("");
    return putObjectToSystemMng(path, body);
  }

  /**
   * Requests a "POST Component".
   * @param body a property object.
   * @return response object.
   */
  public final Response postComponent(
      final ObjectProperty body) {
    String path = COMPS_PATH;
    log.debug("");
    return postObjectToSystemMng(path, body);
  }

  /**
   * Requests a "POST Connection".
   * @param body a connection.
   * @return response object.
   */
  public final Response postConnection(
      final ComponentConnection body) {
    String path = CONNECTIONS_PATH;
    log.debug("");
    return postObjectToSystemMng(path, body);
  }

  /**
   * Requests a "DELETE Component Manager".
   * @param compMngId component manager ID.
   * @return response object.
   */
  public final Response delComponentMng(
      final String compMngId) {
    String path = String.format(COMP_MNG_PATH, compMngId);
    log.debug("");
    return delObjectToSystemMng(path);
  }

  /**
   * Requests a "DELETE Component".
   * @param compId component ID.
   * @return response object.
   */
  public final Response delComponent(
      final String compId) {
    String path = String.format(COMP_PATH, compId);
    log.debug("");
    return delObjectToSystemMng(path);
  }

  /**
   * Requests a "DELETE Connection".
   * @param connId connection ID.
   * @return response object.
   */
  public final Response delConnection(
      final String connId) {
    String path = String.format(CONNECTION_PATH, connId);
    log.debug("");
    return delObjectToSystemMng(path);
  }

  // //////////////////////////////////////////////////
  //
  // common method
  //
  // //////////////////////////////////////////////////

  /**
   * Requests POST to system manager.
   * @param path a path.
   * @param body a resource.
   * @return response object.
   */
  public Response postObjectToSystemMng(
      final String path, final Object body) {
    log.debug("");

    try {
      Response resp = sendRequest(Request.Method.POST, path, body);
      if (resp.isError("POST")) {
        log.warn("invalid POST:" + resp.statusCode);
      }
      return resp;
    } catch (Exception e) {
      log.error("Recieved Message Exception.", e);
      return new Response(Response.INTERNAL_SERVER_ERROR, null);
    }
  }

  /**
   * Requests PUT to system manager.
   * @param path a path.
   * @param body a resource.
   * @return response object.
   */
  public Response putObjectToSystemMng(
      final String path, final Object body) {
    log.debug("");

    try {
      Response resp = sendRequest(Request.Method.PUT, path, body);
      if (resp.isError("PUT")) {
        log.warn("invalid PUT:" + resp.statusCode);
      }
      return resp;
    } catch (Exception e) {
      log.error("Recieved Message Exception.", e);
      return new Response(Response.INTERNAL_SERVER_ERROR, null);
    }
  }

  /**
   * Requests DELETE to system manager.
   * @param path a path.
   * @return response object.
   */
  public Response delObjectToSystemMng(final String path) {
    log.debug("");

    try {
      Response resp = sendRequest(Request.Method.DELETE, path, null);
      if (resp.isError("DELETE")) {
        log.warn("invalid DELETE:" + resp.statusCode);
      }
      return resp;
    } catch (Exception e) {
      log.error("Recieved Message Exception.", e);
      return new Response(Response.INTERNAL_SERVER_ERROR, null);
    }
  }

  /**
   * Requests GET to system manager.
   * @param path a path.
   * @return response object.
   */
  public Response getObjectToSystemMng(String path) {
    log.debug("");

    try {
      Response resp = sendRequest(Request.Method.GET, path, null);
      if (resp.isError("GET")) {
        log.warn("invalid GET:" + resp.statusCode);
      }
      return resp;
    } catch (Exception e) {
      log.error("Recieved Message Exception.", e);
      return new Response(Response.INTERNAL_SERVER_ERROR, null);
    }
  }

  private Response sendRequest(
      final Request.Method method, final String path, final Object body) {
    log.debug("");

    Response rsp = null;
    Request req = new Request(
        this.dispatcher.getSystemManagerId(), method, path, body);
    log.debug("   " + req.getBodyValue());
    try {
      rsp = this.dispatcher.requestSync(req, sourceObjectId);

    } catch (Exception e) {
      log.error("Recieved Message Exception.", e);
      return new Response(Response.INTERNAL_SERVER_ERROR, null);
    }
    return rsp;
  }

}

File: src/main/ruby/org/o3project/odenos/core/component/system_manager_interface.rb

# Copyright 2015 NEC Corporation.                                          #
#                                                                          #
# Licensed under the Apache License, Version 2.0 (the "License");          #
# you may not use this file except in compliance with the License.         #
# You may obtain a copy of the License at                                  #
#                                                                          #
#   http://www.apache.org/licenses/LICENSE-2.0                             #
#                                                                          #
# Unless required by applicable law or agreed to in writing, software      #
# distributed under the License is distributed on an "AS IS" BASIS,        #
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. #
# See the License for the specific language governing permissions and      #
# limitations under the License.                                           #

require 'odenos/remoteobject'
require 'odenos/core/util/logger'

module Odenos
  module Component
    class SystemManagerInterface
      include Odenos::Core
      include Odenos::Util::Logger

      COMP_MNGS_PATH = 'component_managers'
      COMP_MNG_PATH = 'component_managers/%s'
      EVENT_MNG_PATH = 'event_manager'
      COMP_TYPES_PATH = 'component_types'
      COMP_TYPE_PATH = 'component_types/%s'
      COMPS_PATH = 'components'
      COMP_PATH = 'components/%s'
      CONNECTIONS_PATH = 'connections'
      CONNECTION_PATH = 'connections/%s'
      OBJECT_PATH = 'objects/%s'

      def initialize(dispatcher)
        @dispatcher = dispatcher
        @sysmgr_id = dispatcher.system_manager_id
        logger_ident_initialize('SystemManagerInterface')
      end

      ####################
      # request method
      ####################

      # GET Component Managers.
      def get_component_managers
        debug 'GET Component Managers.'
        path = COMP_MNGS_PATH
        resp = get_object_to_sysmgr(path)
        unless resp.status_code == Response::OK
          return nil
        end
        begin
          comp_mgrs = {}
          resp.body.each do |prop|
            objp = ObjectProperty.new(prop)
            comp_mgrs[objp.remote_object_id] = objp
          end
          return comp_mgrs
        rescue => ex
          return dump_error(ex)
        end
      end

      # GET Event Manager.
      def get_event_manager
        debug 'GET Event Manager.'
        path = EVENT_MNG_PATH
        resp = get_object_to_sysmgr(path)
        unless resp.status_code == Response::OK
          return nil
        end
        begin
          return ObjectProperty.new(resp.body)
        rescue => ex
          return dump_error(ex)
        end
      end

      # GET ComponentTypes.
      def get_component_types
        debug 'GET ComponentTypes.'
        path = COMP_TYPES_PATH
        resp = get_object_to_sysmgr(path)
        unless resp.status_code == Response::OK
          return nil
        end
        begin
          comp_types = {}
          resp.body.each do |type, list|
            obj_list = []
            list.each do |prop|
              obj_list.push(ObjectProperty.new(prop))
            end
            comp_types[type] = obj_list
          end
          return comp_types
        rescue => ex
          return dump_error(ex)
        end
      end

      # GET Components.
      def get_components
        debug 'GET Components.'
        path = COMPS_PATH
        resp = get_object_to_sysmgr(path)
        unless resp.status_code == Response::OK
          return nil
        end
        begin
          comps = {}
          resp.body.each do |id, prop|
            objp = ObjectProperty.new(prop)
            comps[id] = objp
          end
          return comps
        rescue => ex
          return dump_error(ex)
        end
      end

      # GET Connections.
      def get_connections
        debug 'GET Connections.'
        path = CONNECTIONS_PATH
        resp = get_object_to_sysmgr(path)
        unless resp.status_code == Response::OK
          return nil
        end
        begin
          conns = {}
          resp.body.each do |id, conn|
            conn_prop = nil
            if conn[:type] == 'LogicAndNetwork'
              conn_prop = LogicAndNetwork.new(conn)
            else
              conn_prop = ComponentConnection.new(conn)
            end
            conns[id] = conn_prop
          end
          return conns
        rescue => ex
          return dump_error(ex)
        end
      end

      # GET Component Manager.
      def get_component_manager(comp_mgr_id)
        debug 'GET Component Manager.'
        path = COMP_MNG_PATH % comp_mgr_id
        resp = get_object_to_sysmgr(path)
        unless resp.status_code == Response::OK
          return nil
        end
        begin
          return ObjectProperty.new(resp.body)
        rescue => ex
          return dump_error(ex)
        end
      end

      # GET ComponentType.
      def get_component_type(comp_type)
        debug 'GET ComponentType.'
        path = COMP_TYPE_PATH % comp_type
        resp = get_object_to_sysmgr(path)
        unless resp.status_code == Response::OK
          return nil
        end
        begin
          obj_list = []
          resp.body.each do |prop|
            obj_list.push(ObjectProperty.new(prop))
          end
          return obj_list
        rescue => ex
          return dump_error(ex)
        end
      end

      # GET Component.
      def get_component(comp_id)
        debug 'GET Component.'
        path = COMP_PATH % comp_id
        resp = get_object_to_sysmgr(path)
        unless resp.status_code == Response::OK
          return nil
        end
        begin
          return ObjectProperty.new(resp.body)
        rescue => ex
          return dump_error(ex)
        end
      end

      # GET Connection.
      def get_connection(conn_id)
        debug 'GET Connection.'
        path = CONNECTION_PATH % conn_id
        resp = get_object_to_sysmgr(path)
        unless resp.status_code == Response::OK
          return nil
        end
        begin
          conn_prop = nil
          if resp.body[:type] == 'LogicAndNetwork'
            conn_prop = LogicAndNetwork.new(resp.body)
          else
            conn_prop = ComponentConnection.new(resp.body)
          end
          return conn_prop
        rescue => ex
          return dump_error(ex)
        end
      end

      # GET Object.
      def get_object(object_id)
        debug 'GET Object.'
        path = OBJECT_PATH % object_id
        resp = get_object_to_sysmgr(path)
        unless resp.status_code == Response::OK
          return nil
        end
        begin
          return ObjectProperty.new(resp.body)
        rescue => ex
          return dump_error(ex)
        end
      end

      # PUT Connection.
      def put_connection(connection)
        debug 'PUT Connection.'
        path = CONNECTION_PATH % connection.component_connection_id
        put_object_to_sysmgr(path, connection)
      end

      # PUT ComponentManagers.
      def put_component_managers(property)
        debug 'PUT ComponentManagers.'
        path = COMP_MNG_PATH % property.remote_object_id
        put_object_to_sysmgr(path, property)
      end

      # PUT Components.
      def put_components(property)
        debug 'PUT Components.'
        path = COMP_PATH % property.remote_object_id
        put_object_to_sysmgr(path, property)
      end

      # POST Components.
      def post_components(property)
        debug 'POST Components.'
        path = COMPS_PATH
        post_object_to_sysmgr(path, property)
      end

      # POST Connections.
      def post_connections(connection)
        debug 'POST Components.'
        path = CONNECTIONS_PATH
        post_object_to_sysmgr(path, connection)
      end

      # DELETE ComponentManagers.
      def del_component_managers(comp_mgr_id)
        debug 'DELETE ComponentManagers.'
        path = COMP_MNG_PATH % comp_mgr_id
        del_object_to_sysmgr(path)
      end

      # DELETE Components.
      def del_components(comp_id)
        debug 'DELETE Components.'
        path = COMP_PATH % comp_id
        del_object_to_sysmgr(path)
      end

      # DELETE Connections.
      def del_connections(conn_id)
        debug 'DELETE Connections.'
        path = CONNECTION_PATH % conn_id
        del_object_to_sysmgr(path)
      end

      ####################
      # common method
      ####################

      def post_object_to_sysmgr(path, body)
        debug ">> sysmgr_id: #{@sysmgr_id} path: #{path}"
        send_request(@sysmgr_id, :POST, path, body)
      end

      def put_object_to_sysmgr(path, body)
        debug ">> sysmgr_id: #{@sysmgr_id} path: #{path}"
        send_request(@sysmgr_id, :PUT, path, body)
      end

      def del_object_to_sysmgr(path)
        debug ">> sysmgr_id: #{@sysmgr_id} path: #{path}"
        send_request(@sysmgr_id, :DELETE, path, nil)
      end

      def get_object_to_sysmgr(path)
        debug ">> sysmgr_id: #{@sysmgr_id} path: #{path}"
        send_request(@sysmgr_id, :GET, path, nil)
      end

      ####################
      # private method
      ####################

      private

      def dump_error(ex)
        error 'Exception: Get Invalid Message'
        error " #{ex.message} #{ex.backtrace}"
        nil
      end

      def send_request(obj_id, method, path, body)
        resp = Response.new(Response::INTERNAL_SERVER_ERROR, nil)
        req = Request.new(obj_id, method, path, body)
        begin
          resp = @dispatcher.request_sync(req)
        rescue => ex
          error "Exception: Request to #{obj_id} Method:#{method} Path:#{path}"
          error " #{ex.message} #{ex.backtrace}"
        end
        resp
      end
    end
  end
end
